### PR TITLE
Add actuators for health and info.

### PR DESCRIPTION
**Pull request by VolkerHartmann**
They show information about disc space and no of entries.

### CODE REVIEW

File: src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java [hunk 1]
```diff
@@ -67,20 +68,21 @@ public class WebSecurityConfig {
...
+            )).permitAll().
+            requestMatchers(EndpointRequest.toAnyEndpoint()). //
+            hasAnyRole("ADMIN", "ACTUATOR"). //
+            antMatchers(HttpMethod.OPTIONS, "/**").
+            permitAll().and().
             sessionManagement().
-            sessionCreationPolicy(SessionCreationPolicy.STATELESS).and();
+            sessionCreationPolicy(SessionCreationPolicy.STATELESS).
+            and()
+            .csrf().disable();
```
Comment (ThomasJejkal) [inline src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java:85]: Keeping this, CSRF can never be enabled. So why keeping the property check afterwards?

File: src/main/java/edu/kit/datamanager/metastore2/health/HealthCheck.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        File repoDir = path.toFile();
+        double total = (double) repoDir.getTotalSpace();
+        double free = (double) repoDir.getFreeSpace();
+        totalSpace = String.format("%.2f GB", total / 1073741824);
+        freeSpace = String.format("%.2f GB (%.0f%%)", (double) repoDir.getFreeSpace() / 1073741824, free * 100.0 / total);
+        properties.put("Total space", totalSpace);
+        properties.put("Free space", freeSpace);
+
+      } catch (Throwable t) {
+        LOG.error("Failed to check repository folder at '" + path.toString() + "'. Returning negative health status.", t);
```
Comment (ThomasJejkal) [inline src/main/java/edu/kit/datamanager/metastore2/health/HealthCheck.java:68]: I would rather remove "Returning negative health status." at this point, as nothing negative is returned here. This comes in the extending class.

File: src/main/java/edu/kit/datamanager/metastore2/health/MetadataRepoHealthCheck.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    if (details.isEmpty()) {
+      return Health.down().withDetail("No of metadata documents", 0).build();
+    } else {
+      details.put("No of metadata documents", Long.toString(metadataRecordDao.count()));
+      return Health.up().withDetails(details).build();
+    }
+  }
+
+  @Override
+  public void contribute(Info.Builder builder) {
```
Comment (ThomasJejkal) [inline src/main/java/edu/kit/datamanager/metastore2/health/MetadataRepoHealthCheck.java:77]: Why putting the same information into the InfoActuator?

File: src/main/java/edu/kit/datamanager/metastore2/health/SchemaRepoHealthCheck.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  @Override
+  public Health health() {
+    LOG.trace("Check for SchemaRepo health information...");
+
+    URL basePath = schemaConfig.getBasepath();
+    Map<String, String> details = testDirectory(basePath);
+
+    if (details.isEmpty()) {
+      return Health.down().withDetail("No of metadata documents", 0).build();
```
Comment (ThomasJejkal) [inline src/main/java/edu/kit/datamanager/metastore2/health/SchemaRepoHealthCheck.java:70]: C&P error....must be "No of schema documents"

**Review by ThomasJejkal — CHANGES_REQUESTED** [review]

File: src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java (hunk 1 above)
Comment (VolkerHartmann) [reply src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java:85]: Just a C&P error. :-(

File: src/main/java/edu/kit/datamanager/metastore2/health/MetadataRepoHealthCheck.java (hunk 3 above)
Comment (VolkerHartmann) [reply src/main/java/edu/kit/datamanager/metastore2/health/MetadataRepoHealthCheck.java:77]: This is meant to be a basis for discussion.
I think it's good to see this information in the /actuator/info section. 
Not absolutely necessary for /actuator/health, but if you already have it...

File: src/main/java/edu/kit/datamanager/metastore2/health/SchemaRepoHealthCheck.java (hunk 4 above)
Comment (VolkerHartmann) [reply src/main/java/edu/kit/datamanager/metastore2/health/SchemaRepoHealthCheck.java:70]: Oops, I missed that. Thank you!

File: src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java [hunk 5]
```diff
@@ -67,18 +68,17 @@ public class WebSecurityConfig {
...
-            .antMatchers(HttpMethod.OPTIONS, "/**").
-            permitAll().
-            and().
+  @Override
+  protected void configure(HttpSecurity http) throws Exception {
+    HttpSecurity httpSecurity = http.authorizeRequests().
+            requestMatchers(EndpointRequest.to(
+                    InfoEndpoint.class,
+                    HealthEndpoint.class
+            )).permitAll().
```
Comment (Pfeil) [inline src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java:77]: I might be wrong, but it reads like you give permissions to everyone for the health+info endpoints and then restrict it to ROLE_ACTUATOR (and role_admin) for all other endpoints. Shouldn't it be the other way around?

Comment (ThomasJejkal) [reply src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java:77]: I think the reason is, that it's a filter chain which is left as soon as one rule matches. Currently, I suppose that everybody can access info and health, health details only when authenticated, and all other endpoints of having one of the special roles. Seems fine to me.

Comment (ThomasJejkal) [reply src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java:77]: The first point is only visible in application.properties. in principle the order should be as follows:

* everybody sees info and basis health
* authenticated users see in addition complete health
* special roles also see other actuators, e.g., beans, loggers, configuration properties

Does this make sense?

Comment (Pfeil) [reply src/main/java/edu/kit/datamanager/metastore2/configuration/WebSecurityConfig.java:78]: It does. I also found the management settings in the config file, thanks. It's a bit confusing it is not in the same place, but I see the idea behind it. There is still one point of confusion. Right below this line, it is like this:

```java
.requestMatchers(EndpointRequest.toAnyEndpoint())
.hasAnyRole("ADMIN", "ACTUATOR")
```

But wouldn't this also affect the actual application endpoints / APIs? Is being admin/actuator required to use the metastore API, and why do actuator roles need access?

**Pull request merged**